File: sdks/react-native/src/storage.ts
import type { ApplAdClient } from './client';

export class Storage {
  constructor(private client: ApplAdClient) {}

  // --- Buckets ---

  createBucket(
    name: string,
    opts?: {
      bucketId?: string;
      permissions?: string[];
      maximumFileSize?: number;
      allowedFileExtensions?: string[];
      compression?: string;
      encryption?: boolean;
      antivirus?: boolean;
    }
  ) {
    return this.client.call('POST', '/storage/buckets', {
      name,
      bucketId: opts?.bucketId ?? 'unique()',
      permissions: opts?.permissions ?? [],
      allowedFileExtensions: opts?.allowedFileExtensions ?? [],
      ...(opts?.maximumFileSize && { maximumFileSize: opts.maximumFileSize }),
      ...(opts?.compression && { compression: opts.compression }),
      encryption: opts?.encryption ?? false,
      antivirus: opts?.antivirus ?? false,
    });
  }

  listBuckets() {
    return this.client.call('GET', '/storage/buckets');
  }

  getBucket(bucketId: string) {
    return this.client.call('GET', `/storage/buckets/${bucketId}`);
  }

  updateBucket(
    bucketId: string,
    name: string,
    opts?: { permissions?: string[]; maximumFileSize?: number; enabled?: boolean }
  ) {
    return this.client.call('PUT', `/storage/buckets/${bucketId}`, {
      name,
      ...(opts?.permissions && { permissions: opts.permissions }),
      ...(opts?.maximumFileSize && { maximumFileSize: opts.maximumFileSize }),
      ...(opts?.enabled !== undefined && { enabled: opts.enabled }),
    });
  }

  deleteBucket(bucketId: string) {
    return this.client.call('DELETE', `/storage/buckets/${bucketId}`);
  }

  // --- Files ---

  /**
   * Upload a file. Accepts React Native compatible file descriptors.
   *
   * @param bucketId - Target bucket ID
   * @param fileUri - Local file URI (e.g. from image picker: "file:///path/to/image.jpg")
   * @param fileName - The file name to use
   * @param mimeType - MIME type (e.g. "image/jpeg")
   * @param opts - Optional file ID and permissions
   */
  uploadFile(
    bucketId: string,
    fileUri: string,
    fileName: string,
    mimeType: string,
    opts?: { fileId?: string; permissions?: string[] }
  ) {
    const formData = new FormData();
    formData.append('fileId', opts?.fileId ?? 'unique()');
    // React Native FormData accepts { uri, name, type } objects
    formData.append('file', {
      uri: fileUri,
      name: fileName,
      type: mimeType,
    } as any);
    if (opts?.permissions) {
      formData.append('permissions', JSON.stringify(opts.permissions));
    }
    return this.client.upload(`/storage/buckets/${bucketId}/files`, formData);
  }

  listFiles(bucketId: string, opts?: { limit?: number; offset?: number }) {
    const params = new URLSearchParams();
    if (opts?.limit) params.set('limit', String(opts.limit));
    if (opts?.offset) params.set('offset', String(opts.offset));
    const qs = params.toString();
    return this.client.call('GET', `/storage/buckets/${bucketId}/files${qs ? `?${qs}` : ''}`);
  }

  getFile(bucketId: string, fileId: string) {
    return this.client.call('GET', `/storage/buckets/${bucketId}/files/${fileId}`);
  }

  deleteFile(bucketId: string, fileId: string) {
    return this.client.call('DELETE', `/storage/buckets/${bucketId}/files/${fileId}`);
  }

  downloadFile(bucketId: string, fileId: string) {
    return this.client.download(`/storage/buckets/${bucketId}/files/${fileId}/download`);
  }

  /**
   * Build a URL for file preview/thumbnail with optional transforms.
   */
  getFilePreview(
    bucketId: string,
    fileId: string,
    opts?: { width?: number; height?: number; quality?: number; format?: string }
  ): string {
    const params = new URLSearchParams();
    params.set('project', this.client.projectId);
    if (opts?.width) params.set('width', String(opts.width));
    if (opts?.height) params.set('height', String(opts.height));
    if (opts?.quality) params.set('quality', String(opts.quality));
    if (opts?.format) params.set('output', opts.format);
    return `${this.client.endpoint}/v1/storage/buckets/${bucketId}/files/${fileId}/preview?${params.toString()}`;
  }
}

File: sdks/dart/lib/src/server/storage.dart
import '../server_client.dart';

/// Storage service — manage buckets and files (server-side).
class Storage {
  final ApplAdServer _client;

  Storage(this._client);

  Future<Map<String, dynamic>> createBucket({
    required String name,
    String? bucketId,
    List<String>? permissions,
    int? maximumFileSize,
    List<String>? allowedFileExtensions,
    String? compression,
    bool encryption = false,
    bool antivirus = false,
  }) async {
    return _client.post('/v1/storage/buckets', data: {
      'name': name,
      'bucketId': bucketId ?? 'unique()',
      'permissions': permissions ?? [],
      if (maximumFileSize != null) 'maximumFileSize': maximumFileSize,
      'allowedFileExtensions': allowedFileExtensions ?? [],
      if (compression != null) 'compression': compression,
      'encryption': encryption,
      'antivirus': antivirus,
    });
  }

  Future<Map<String, dynamic>> listBuckets() async {
    return _client.get('/v1/storage/buckets');
  }

  Future<Map<String, dynamic>> getBucket(String bucketId) async {
    return _client.get('/v1/storage/buckets/$bucketId');
  }

  Future<Map<String, dynamic>> updateBucket({
    required String bucketId,
    required String name,
    List<String>? permissions,
    int? maximumFileSize,
    bool? enabled,
  }) async {
    return _client.put('/v1/storage/buckets/$bucketId', data: {
      'name': name,
      if (permissions != null) 'permissions': permissions,
      if (maximumFileSize != null) 'maximumFileSize': maximumFileSize,
      if (enabled != null) 'enabled': enabled,
    });
  }

  Future<Map<String, dynamic>> deleteBucket(String bucketId) async {
    return _client.delete('/v1/storage/buckets/$bucketId');
  }

  Future<Map<String, dynamic>> createFile({
    required String bucketId,
    required String fileName,
    String? fileId,
    List<String>? permissions,
  }) async {
    return _client.post('/v1/storage/buckets/$bucketId/files', data: {
      'fileId': fileId ?? 'unique()',
      'fileName': fileName,
      if (permissions != null) 'permissions': permissions,
    });
  }

  Future<Map<String, dynamic>> listFiles({
    required String bucketId,
    int? limit,
    int? offset,
  }) async {
    final query = <String, String>{};
    if (limit != null) query['limit'] = limit.toString();
    if (offset != null) query['offset'] = offset.toString();
    final qs = query.isNotEmpty
        ? '?${query.entries.map((e) => '${e.key}=${Uri.encodeComponent(e.value)}').join('&')}'
        : '';
    return _client.get('/v1/storage/buckets/$bucketId/files$qs');
  }

  Future<Map<String, dynamic>> getFile({
    required String bucketId,
    required String fileId,
  }) async {
    return _client.get('/v1/storage/buckets/$bucketId/files/$fileId');
  }

  Future<Map<String, dynamic>> deleteFile({
    required String bucketId,
    required String fileId,
  }) async {
    return _client.delete('/v1/storage/buckets/$bucketId/files/$fileId');
  }
}
